tiledb/sm/rest/rest_profile.cc

```diff
@@ -0,0 +1,289 @@
+/*       PARAMETER DEFAULTS       */
+/* ****************************** */
+
+const std::string RestProfile::DEFAULT_NAME{"default"};
+const std::string RestProfile::DEFAULT_PASSWORD{""};
+const std::string RestProfile::DEFAULT_PAYER_NAMESPACE{""};
+const std::string RestProfile::DEFAULT_TOKEN{""};
+const std::string RestProfile::DEFAULT_SERVER_ADDRESS{"https://api.tiledb.com"};
+const std::string RestProfile::DEFAULT_USERNAME{""};
+
+const std::map<std::string, std::string> default_values = {
+    std::make_pair("rest.password", RestProfile::DEFAULT_PASSWORD),
+    std::make_pair(
+        "rest.payer_namespace", RestProfile::DEFAULT_PAYER_NAMESPACE),
+    std::make_pair("rest.token", RestProfile::DEFAULT_TOKEN),
+    std::make_pair("rest.server_address", RestProfile::DEFAULT_SERVER_ADDRESS),
+    std::make_pair("rest.username", RestProfile::DEFAULT_USERNAME)};
+
+/* ****************************** */
+/*   CONSTRUCTORS & DESTRUCTORS   */
+/* ****************************** */
+
+RestProfile::RestProfile(const std::string& name, const std::string& homedir)
+    : version_(constants::rest_profile_version)
+    , name_(name)
+    , filepath_(homedir + ".tiledb/profiles.json")
+    , old_filepath_(homedir + ".tiledb/cloud.json")
+    , param_values_(default_values) {
+  /**
+   * Ensure the user's $HOME is found.
+   * There's an edge case in which `sudo` does not always preserve the path to
+   * `$HOME`. In this case, the home_directory() API does not throw, but instead
+   * returns `std::nullopt`. As such, we can check for a value in the returned
+   * `std::optional` path of the home_directory and throw an error to the user
+   * accordingly, so they may decide the proper course of action: set the
+   * $HOME path, or perhaps stop using `sudo`.
+   */
+  if (homedir.empty()) {
+    throw RestProfileException(
+        "Failed to create RestProfile; $HOME is not set.");
+  }
+
+  // Fstream cannot create directories. If `homedir/.tiledb/` DNE, create it.
+  std::filesystem::create_directories(homedir + ".tiledb");
+
+  // If the local file exists, load the profile with the given name.
+  if (std::filesystem::exists(filepath_)) {
+    load_from_json_file(filepath_);
+  } else {
+    // If the old version of the file exists, load the profile from there
+    if (std::filesystem::exists(old_filepath_)) {
+      load_from_json_file(old_filepath_);
+    }
+  }
+}
+
+RestProfile::RestProfile(const std::string& name)
+    : RestProfile(
+          name, home_directory().has_value() ? home_directory().value() : "") {
+}
+
+/* ****************************** */
+/*              API               */
+/* ****************************** */
+
+void RestProfile::set(const std::string& param, const std::string& value) {
+  // Validate incoming parameter name
+  if (default_values.count(param) == 0) {
+    throw RestProfileException(
+        "Failed to set parameter of invalid name \'" + param + "\'");
+  }
+  param_values_[param] = value;
```

Comment:
Do we allow to set parameter names that are already set to non-default values? I remember we discussed about profiles being immutable.

Reply:
I see now that we check if the name of the profile exists before saving, so never mind here.